src/MongoDB.Driver.Core/Core/Compression/Snappy/SnappyCodec.cs

```diff
@@ -0,0 +1,166 @@
+            {
+                case SnappyStatus.Ok:
+                    return outputLength;
+                default:
+                    throw new InvalidDataException("Input is not a valid snappy-compressed block.");
+            }
+        }
+
+        public static int GetUncompressedLength(byte[] input)
+        {
+            Ensure.IsNotNull(input, nameof(input));
+
+            return GetUncompressedLength(input, 0, input.Length);
+        }
+
+        public static int Uncompress(byte[] input, int inputOffset, int inputLength, byte[] output, int outputOffset, int outputLength)
+        {
+            Ensure.IsNotNull(input, nameof(input));
+            Ensure.IsNotNull(output, nameof(output));
+            EnsureBufferRangeIsValid(inputOffset, inputLength, input.Length);
+            EnsureBufferRangeIsValid(outputOffset, outputLength, output.Length);
+
+            if (inputLength == 0)
+            {
+                throw new InvalidDataException("Compressed block cannot be empty.");
+            }
+
+            if (outputOffset == output.Length)
+            {
+                output = new byte[1];
+                outputOffset = 0;
+            }
+
+            var status = SnappyAdapter.snappy_uncompress(input, inputOffset, inputLength, output, outputOffset, ref outputLength);
+            switch (status)
+            {
+                case SnappyStatus.Ok:
+                    return outputLength;
+                case SnappyStatus.BufferTooSmall:
+                    throw new ArgumentOutOfRangeException("Output array is too small.");
+                default:
+                    throw new InvalidDataException("Input is not a valid snappy-compressed block.");
+            }
+        }
+
+        public static byte[] Uncompress(byte[] input)
+        {
+            var max = GetUncompressedLength(input);
+            var output = new byte[max];
+            var outputLength = Uncompress(input, 0, input.Length, output, 0, output.Length);
+            if (outputLength == max)
+                return output;
+
+            var truncated = new byte[outputLength];
+            Array.Copy(output, truncated, outputLength);
+            return truncated;
+        }
+
+        public static bool Validate(byte[] input, int inputOffset, int inputLength)
+        {
+            Ensure.IsNotNull(input, nameof(input));
+            EnsureBufferRangeIsValid(inputOffset, inputLength, input.Length);
+            if (inputLength == 0)
+            {
+                return false;
+            }
+
+            return SnappyAdapter.snappy_validate_compressed_buffer(input, inputOffset, inputLength) == SnappyStatus.Ok;
+        }
+
+        public static bool Validate(byte[] input)
+        {
+            Ensure.IsNotNull(input, nameof(input));
+
+            return Validate(input, 0, input.Length);
+        }
+
+        // private static methods
+        private static void EnsureBufferRangeIsValid(int offset, int length, int bufferLength)
```

Comment:
Should be double-checked

Reply:
Checked. This validation is not equivalent to my originally implemented. For example, this test https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/82/files#diff-499895a7c691ac6fb1ae1212ae89cb29R57  will return the same result (`ArgumentOutOfRangeException`) but from different code lines. The suggested method will ignore this test condition and the error will be thrown only inside the native library. 

However, It will affect only output validation in `Compress` method and the cases when `offset` is the same as `bufferLength`. Since our `SnappyCompressor` implementation passes only `outputOffset` equal to `0`, the suggested change won't affect our code. Also, it looks like even if it will be changed someday, it looks like we will still be able to see the expected error, it will be just thrown from the native library.

So, I don't mind about the suggested change here.

Reply:
cc: @rstam

Reply:
This validation is not intended to be equivalent to any prior implementation.

It's just intended to validate that `offset`, `length`, and `bufferLength` represent a valid non-zero length range of a bugger.